plugins/samples/domain_rewrite/plugin.rs

```diff
@@ -0,0 +1,103 @@
+}}
+
+struct MyHttpContext<'a> {
+    // Some member variables need to be wrapped in Rc<Refcell<T>> so that we can
+    // modify the variable in multiple places as well as share ownership. By
+    // using Rc, we are commiting to not modfying the variable in multiple places
+    // simultaneously, otherwise the code will panic and plugin will crash.
+
+    // Stores HTML as rewriter parses and modifies HTML.
+    // Only stores completed sections(e.g., when rewriter parses "<di", output
+    // will be empty until the next chunk comes in. If the next chunk was
+    // "v> <h1>foo</h", the output will then contain "<div><h1>foo").
+    output: Rc<RefCell<Vec<u8>>>,
+    // HTML rewriter. Member of MyHttpContext a.k.a "StreamContext" so that the
+    // rewriter persists across multiple body callbacks.
+    rewriter: Option<HtmlRewriter<'a, Box<dyn FnMut(&[u8])>>>,
+}
+
+impl<'a> MyHttpContext<'a> {
+    pub fn new() -> Box<MyHttpContext<'a>> {
+        let output = Rc::new(RefCell::new(Vec::new()));
+        Box::new(MyHttpContext {
+            output: output.clone(),
+            rewriter: Some(HtmlRewriter::new(
+                Settings {
+                    element_content_handlers: vec![element!("a[href]", move |el| {
+                        let href = el.get_attribute("href").unwrap();
+                        let modified_href = href.replace("foo", "bar");
+                        if modified_href != href {
+                            el.set_attribute("href", &modified_href).unwrap();
+                        }
+
+                        Ok(())
+                    })],
+                    ..Settings::new()
+                },
+                Box::new(move |c: &[u8]| output.borrow_mut().extend_from_slice(c)),
+            )),
+        })
+    }
+
+    fn parse_chunk(&mut self, body_bytes: Bytes) -> Result<(), Box<dyn Error>> {
+        // Parse/rewrite current chunk
+        self.rewriter
+            .as_mut()
+            .ok_or("Expected valid rewriter. Got None")?
+            .write(&body_bytes)?;
+        return Ok(());
+    }
+}
+
+impl<'a> Context for MyHttpContext<'a> {}
+
+impl<'a> HttpContext for MyHttpContext<'a> {
+    fn on_http_request_headers(&mut self, _: usize, _: bool) -> Action {
+        return Action::Continue;
+    }
+
+    fn on_http_response_body(&mut self, body_size: usize, _: bool) -> Action {
+        if let Some(body_bytes) = self.get_http_response_body(0, body_size) {
+            if let Err(e) = self.parse_chunk(body_bytes) {
+                self.send_http_response(
+                    500,
+                    vec![],
+                    Some(
+                        &format!("Error while writing to HtmlRewriter: {}", e.to_string())
+                            .into_bytes(),
+                    ),
+                );
+            }
+        }
+        self.set_http_response_body(0, body_size, self.output.borrow().as_slice());
```

Comment:
Pretty sure the `body_size` param should instead be the size of the data we're about to send back (the number of bytes in self.output).

Reply:
Oh wow, so this is the body content to replace. Can you add a quick comment to call this out. For example:

`// Replace the entire chunk (0 to body_size) with the latest data emitted by the rewriter`